File: dart/packages/fory/lib/src/serializer/primitive_serializers.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import 'package:fory/src/context/read_context.dart';
import 'package:fory/src/context/write_context.dart';
import 'package:fory/src/meta/type_ids.dart';
import 'package:fory/src/serializer/serializer.dart';
import 'package:fory/src/types/float32.dart';
import 'package:fory/src/types/int64.dart';
import 'package:fory/src/types/uint64.dart';
import 'package:fory/src/util/int_validation.dart';

const int _jsSafeUint64IntMax = 9007199254740991;
const bool _isWeb = bool.fromEnvironment('dart.library.js_interop');

final class PrimitiveSerializer<T> extends Serializer<T> {
  final int typeId;
  final bool _supportsRef;

  const PrimitiveSerializer(
    this.typeId, {
    required bool supportsRef,
  }) : _supportsRef = supportsRef;

  @override
  bool get supportsRef => _supportsRef;

  @override
  void write(WriteContext context, T value) {
    writePayload(context, typeId, value as Object);
  }

  @override
  T read(ReadContext context) {
    return readPayload(context, typeId) as T;
  }

  static void writePayload(
    WriteContext context,
    int typeId,
    Object value,
  ) {
    final buffer = context.buffer;
    switch (typeId) {
      case TypeIds.boolType:
        buffer.writeBool(value as bool);
        return;
      case TypeIds.int8:
        buffer.writeByte(checkedInt8(value as int));
        return;
      case TypeIds.int16:
        buffer.writeInt16(checkedInt16(value as int));
        return;
      case TypeIds.int32:
        buffer.writeInt32(checkedInt32(value as int));
        return;
      case TypeIds.varInt32:
        buffer.writeVarInt32(checkedInt32(value as int));
        return;
      case TypeIds.int64:
        if (value is Int64) {
          buffer.writeInt64(value);
        } else {
          buffer.writeInt64FromInt(value as int);
        }
        return;
      case TypeIds.varInt64:
        if (value is Int64) {
          buffer.writeVarInt64(value);
        } else {
          buffer.writeVarInt64FromInt(value as int);
        }
        return;
      case TypeIds.taggedInt64:
        if (value is Int64) {
          buffer.writeTaggedInt64(value);
        } else {
          buffer.writeTaggedInt64FromInt(value as int);
        }
        return;
      case TypeIds.uint8:
        buffer.writeUint8(checkedUint8(value as int));
        return;
      case TypeIds.uint16:
        buffer.writeUint16(checkedUint16(value as int));
        return;
      case TypeIds.uint32:
        buffer.writeUint32(checkedUint32(value as int));
        return;
      case TypeIds.varUint32:
        buffer.writeVarUint32(checkedUint32(value as int));
        return;
      case TypeIds.uint64:
        buffer.writeUint64(_uint64Value(value));
        return;
      case TypeIds.varUint64:
        buffer.writeVarUint64(_uint64Value(value));
        return;
      case TypeIds.taggedUint64:
        buffer.writeTaggedUint64(_uint64Value(value));
        return;
      case TypeIds.float16:
        buffer.writeFloat16(value as double);
        return;
      case TypeIds.bfloat16:
        buffer.writeBfloat16(value as double);
        return;
      case TypeIds.float32:
        buffer.writeFloat32((value as Float32).value);
        return;
      case TypeIds.float64:
        buffer.writeFloat64(value as double);
        return;
      default:
        throw StateError('Unsupported primitive type id $typeId.');
    }
  }

  static Object readPayload(
    ReadContext context,
    int typeId,
  ) {
    final buffer = context.buffer;
    switch (typeId) {
      case TypeIds.boolType:
        return buffer.readBool();
      case TypeIds.int8:
        return buffer.readByte();
      case TypeIds.int16:
        return buffer.readInt16();
      case TypeIds.int32:
        return buffer.readInt32();
      case TypeIds.varInt32:
        return buffer.readVarInt32();
      case TypeIds.int64:
        return buffer.readInt64();
      case TypeIds.varInt64:
        return buffer.readVarInt64();
      case TypeIds.taggedInt64:
        return buffer.readTaggedInt64();
      case TypeIds.uint8:
        return buffer.readUint8();
      case TypeIds.uint16:
        return buffer.readUint16();
      case TypeIds.uint32:
        return buffer.readUint32();
      case TypeIds.varUint32:
        return buffer.readVarUint32();
      case TypeIds.uint64:
        return buffer.readUint64();
      case TypeIds.varUint64:
        return buffer.readVarUint64();
      case TypeIds.taggedUint64:
        return buffer.readTaggedUint64();
      case TypeIds.float16:
        return buffer.readFloat16();
      case TypeIds.bfloat16:
        return buffer.readBfloat16();
      case TypeIds.float32:
        return Float32(buffer.readFloat32());
      case TypeIds.float64:
        return buffer.readFloat64();
      default:
        throw StateError('Unsupported primitive type id $typeId.');
    }
  }
}

Uint64 _uint64Value(Object value) {
  if (value is Uint64) {
    return value;
  }
  final intValue = value as int;
  if (_isWeb && (intValue < 0 || intValue > _jsSafeUint64IntMax)) {
    throw StateError(
      'Dart int value $intValue is outside the JS-safe unsigned uint64 '
      'int field range [0, $_jsSafeUint64IntMax]. Use Uint64 for full '
      'unsigned 64-bit values on web.',
    );
  }
  return Uint64(intValue);
}

const PrimitiveSerializer<bool> boolSerializer = PrimitiveSerializer<bool>(
  TypeIds.boolType,
  supportsRef: false,
);
const PrimitiveSerializer<int> int8Serializer = PrimitiveSerializer<int>(
  TypeIds.int8,
  supportsRef: false,
);
const PrimitiveSerializer<int> int16Serializer = PrimitiveSerializer<int>(
  TypeIds.int16,
  supportsRef: false,
);
const PrimitiveSerializer<int> int32Serializer = PrimitiveSerializer<int>(
  TypeIds.int32,
  supportsRef: false,
);
const PrimitiveSerializer<int> varInt32Serializer = PrimitiveSerializer<int>(
  TypeIds.varInt32,
  supportsRef: false,
);
const PrimitiveSerializer<Int64> int64Serializer = PrimitiveSerializer<Int64>(
  TypeIds.int64,
  supportsRef: false,
);
const PrimitiveSerializer<Int64> varInt64Serializer =
    PrimitiveSerializer<Int64>(
  TypeIds.varInt64,
  supportsRef: false,
);
const PrimitiveSerializer<Int64> taggedInt64Serializer =
    PrimitiveSerializer<Int64>(
  TypeIds.taggedInt64,
  supportsRef: false,
);
const PrimitiveSerializer<int> uint8Serializer = PrimitiveSerializer<int>(
  TypeIds.uint8,
  supportsRef: false,
);
const PrimitiveSerializer<int> uint16Serializer = PrimitiveSerializer<int>(
  TypeIds.uint16,
  supportsRef: false,
);
const PrimitiveSerializer<int> uint32Serializer = PrimitiveSerializer<int>(
  TypeIds.uint32,
  supportsRef: false,
);
const PrimitiveSerializer<int> varUint32Serializer = PrimitiveSerializer<int>(
  TypeIds.varUint32,
  supportsRef: false,
);
const PrimitiveSerializer<Uint64> uint64Serializer =
    PrimitiveSerializer<Uint64>(
  TypeIds.uint64,
  supportsRef: false,
);
const PrimitiveSerializer<Uint64> varUint64Serializer =
    PrimitiveSerializer<Uint64>(
  TypeIds.varUint64,
  supportsRef: false,
);
const PrimitiveSerializer<Uint64> taggedUint64Serializer =
    PrimitiveSerializer<Uint64>(
  TypeIds.taggedUint64,
  supportsRef: false,
);
const PrimitiveSerializer<double> float16Serializer =
    PrimitiveSerializer<double>(
  TypeIds.float16,
  supportsRef: false,
);
const PrimitiveSerializer<double> bfloat16Serializer =
    PrimitiveSerializer<double>(
  TypeIds.bfloat16,
  supportsRef: false,
);
const PrimitiveSerializer<Float32> float32Serializer =
    PrimitiveSerializer<Float32>(
  TypeIds.float32,
  supportsRef: false,
);
const PrimitiveSerializer<double> float64Serializer =
    PrimitiveSerializer<double>(
  TypeIds.float64,
  supportsRef: false,
);

File: java/fory-core/src/main/java/org/apache/fory/serializer/PrimitiveSerializers.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.serializer;

import static org.apache.fory.type.TypeUtils.PRIMITIVE_LONG_TYPE;

import org.apache.fory.codegen.Expression;
import org.apache.fory.codegen.Expression.Invoke;
import org.apache.fory.config.Config;
import org.apache.fory.config.Int64Encoding;
import org.apache.fory.context.ReadContext;
import org.apache.fory.context.WriteContext;
import org.apache.fory.memory.MemoryBuffer;
import org.apache.fory.memory.NativeByteOrder;
import org.apache.fory.resolver.TypeResolver;
import org.apache.fory.type.BFloat16;
import org.apache.fory.type.Float16;
import org.apache.fory.util.Preconditions;

/** Serializers for java primitive types. */
@SuppressWarnings({"rawtypes", "unchecked"})
public class PrimitiveSerializers {
  public static final class BooleanSerializer extends ImmutableSerializer<Boolean>
      implements Shareable {
    public BooleanSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Boolean value) {
      writeContext.getBuffer().writeBoolean(value);
    }

    @Override
    public Boolean read(ReadContext readContext) {
      return readContext.getBuffer().readBoolean();
    }
  }

  public static final class ByteSerializer extends ImmutableSerializer<Byte> implements Shareable {
    public ByteSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Byte value) {
      writeContext.getBuffer().writeByte(value);
    }

    @Override
    public Byte read(ReadContext readContext) {
      return readContext.getBuffer().readByte();
    }
  }

  public static final class UInt8Serializer extends ImmutableSerializer<Integer>
      implements Shareable {
    public UInt8Serializer(Config config) {
      super(config, Integer.class);
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      Preconditions.checkArgument(value >= 0 && value <= 255);
      writeContext.getBuffer().writeByte(value.byteValue());
    }

    @Override
    public Integer read(ReadContext readContext) {
      return readContext.getBuffer().readByte() & 0xFF;
    }
  }

  public static final class UInt16Serializer extends ImmutableSerializer<Integer>
      implements Shareable {
    public UInt16Serializer(Config config) {
      super(config, Integer.class);
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      Preconditions.checkArgument(value >= 0 && value <= 65535);
      writeContext.getBuffer().writeInt16(value.shortValue());
    }

    @Override
    public Integer read(ReadContext readContext) {
      return readContext.getBuffer().readInt16() & 0xFFFF;
    }
  }

  public static final class CharSerializer extends ImmutableSerializer<Character>
      implements Shareable {
    public CharSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Character value) {
      writeContext.getBuffer().writeChar(value);
    }

    @Override
    public Character read(ReadContext readContext) {
      return readContext.getBuffer().readChar();
    }
  }

  public static final class ShortSerializer extends ImmutableSerializer<Short>
      implements Shareable {
    public ShortSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Short value) {
      writeContext.getBuffer().writeInt16(value);
    }

    @Override
    public Short read(ReadContext readContext) {
      return readContext.getBuffer().readInt16();
    }
  }

  public static final class IntSerializer extends ImmutableSerializer<Integer>
      implements Shareable {
    private final boolean compressNumber;

    public IntSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
      // Cross-language encoding always uses varint; Java mode follows compressInt config.
      compressNumber = config.isXlang() || config.compressInt();
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      if (compressNumber) {
        writeContext.getBuffer().writeVarInt32(value);
      } else {
        writeContext.getBuffer().writeInt32(value);
      }
    }

    @Override
    public Integer read(ReadContext readContext) {
      if (compressNumber) {
        return readContext.getBuffer().readVarInt32();
      }
      return readContext.getBuffer().readInt32();
    }
  }

  public static final class FixedInt32Serializer extends ImmutableSerializer<Integer>
      implements Shareable {
    public FixedInt32Serializer(Config config) {
      super(config, Integer.class);
    }

    @Override
    public void write(WriteContext writeContext, Integer value) {
      writeContext.getBuffer().writeInt32(value);
    }

    @Override
    public Integer read(ReadContext readContext) {
      return readContext.getBuffer().readInt32();
    }
  }

  public static final class FixedUInt32Serializer extends ImmutableSerializer<Long>
      implements Shareable {
    public FixedUInt32Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      Preconditions.checkArgument((value & 0xffffffff00000000L) == 0);
      writeContext.getBuffer().writeInt32(value.intValue());
    }

    @Override
    public Long read(ReadContext readContext) {
      return Integer.toUnsignedLong(readContext.getBuffer().readInt32());
    }
  }

  public static final class VarUInt32Serializer extends Serializer<Long> implements Shareable {
    public VarUInt32Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      Preconditions.checkArgument((value & 0xffffffff00000000L) == 0);
      writeContext.getBuffer().writeVarUInt32(value.intValue());
    }

    @Override
    public Long read(ReadContext readContext) {
      return Integer.toUnsignedLong(readContext.getBuffer().readVarUInt32());
    }
  }

  public static final class LongSerializer extends ImmutableSerializer<Long> implements Shareable {
    private final Int64Encoding longEncoding;

    public LongSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
      longEncoding = config.isXlang() ? Int64Encoding.VARINT : config.longEncoding();
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeInt64(writeContext.getBuffer(), value, longEncoding);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readInt64(readContext.getBuffer(), longEncoding);
    }

    public static Expression writeInt64(
        Expression buffer, Expression v, Int64Encoding longEncoding, boolean ensureBounds) {
      switch (longEncoding) {
        case FIXED:
          return new Invoke(buffer, "writeInt64", v);
        case TAGGED:
          return new Invoke(
              buffer, ensureBounds ? "writeTaggedInt64" : "_unsafeWriteTaggedInt64", v);
        case VARINT:
          return new Invoke(buffer, ensureBounds ? "writeVarInt64" : "_unsafeWriteVarInt64", v);
        default:
          throw new UnsupportedOperationException("Unsupported long encoding " + longEncoding);
      }
    }

    public static void writeInt64(MemoryBuffer buffer, long value, Int64Encoding longEncoding) {
      if (longEncoding == Int64Encoding.TAGGED) {
        buffer.writeTaggedInt64(value);
      } else if (longEncoding == Int64Encoding.FIXED) {
        buffer.writeInt64(value);
      } else {
        buffer.writeVarInt64(value);
      }
    }

    public static long readInt64(MemoryBuffer buffer, Int64Encoding longEncoding) {
      if (longEncoding == Int64Encoding.TAGGED) {
        return buffer.readTaggedInt64();
      } else if (longEncoding == Int64Encoding.FIXED) {
        return buffer.readInt64();
      } else {
        return buffer.readVarInt64();
      }
    }

    public static Expression readInt64(Expression buffer, Int64Encoding longEncoding) {
      return new Invoke(buffer, readLongFunc(longEncoding), PRIMITIVE_LONG_TYPE);
    }

    public static String readLongFunc(Int64Encoding longEncoding) {
      switch (longEncoding) {
        case FIXED:
          return NativeByteOrder.IS_LITTLE_ENDIAN ? "_readInt64OnLE" : "_readInt64OnBE";
        case TAGGED:
          return NativeByteOrder.IS_LITTLE_ENDIAN ? "_readTaggedInt64OnLE" : "_readTaggedInt64OnBE";
        case VARINT:
          return NativeByteOrder.IS_LITTLE_ENDIAN ? "_readVarInt64OnLE" : "_readVarInt64OnBE";
        default:
          throw new UnsupportedOperationException("Unsupported long encoding " + longEncoding);
      }
    }
  }

  public static final class FixedInt64Serializer extends ImmutableSerializer<Long>
      implements Shareable {
    public FixedInt64Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeContext.getBuffer().writeInt64(value);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readContext.getBuffer().readInt64();
    }
  }

  public static final class TaggedInt64Serializer extends Serializer<Long> implements Shareable {
    public TaggedInt64Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeContext.getBuffer().writeTaggedInt64(value);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readContext.getBuffer().readTaggedInt64();
    }
  }

  public static final class FixedUInt64Serializer extends ImmutableSerializer<Long>
      implements Shareable {
    public FixedUInt64Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeContext.getBuffer().writeInt64(value);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readContext.getBuffer().readInt64();
    }
  }

  public static final class VarUInt64Serializer extends Serializer<Long> implements Shareable {
    public VarUInt64Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeContext.getBuffer().writeVarUInt64(value);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readContext.getBuffer().readVarUInt64();
    }
  }

  public static final class TaggedUInt64Serializer extends Serializer<Long> implements Shareable {
    public TaggedUInt64Serializer(Config config) {
      super(config, Long.class);
    }

    @Override
    public void write(WriteContext writeContext, Long value) {
      writeContext.getBuffer().writeTaggedUInt64(value);
    }

    @Override
    public Long read(ReadContext readContext) {
      return readContext.getBuffer().readTaggedUInt64();
    }
  }

  public static final class FloatSerializer extends ImmutableSerializer<Float>
      implements Shareable {
    public FloatSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Float value) {
      writeContext.getBuffer().writeFloat32(value);
    }

    @Override
    public Float read(ReadContext readContext) {
      return readContext.getBuffer().readFloat32();
    }
  }

  public static final class DoubleSerializer extends ImmutableSerializer<Double>
      implements Shareable {
    public DoubleSerializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Double value) {
      writeContext.getBuffer().writeFloat64(value);
    }

    @Override
    public Double read(ReadContext readContext) {
      return readContext.getBuffer().readFloat64();
    }
  }

  public static final class Float16Serializer extends ImmutableSerializer<Float16>
      implements Shareable {
    public Float16Serializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, Float16 value) {
      writeContext.getBuffer().writeInt16(value.toBits());
    }

    @Override
    public Float16 read(ReadContext readContext) {
      return Float16.fromBits(readContext.getBuffer().readInt16());
    }
  }

  public static final class BFloat16Serializer extends ImmutableSerializer<BFloat16>
      implements Shareable {
    public BFloat16Serializer(Config config, Class<?> cls) {
      super(config, (Class) cls, false);
    }

    @Override
    public void write(WriteContext writeContext, BFloat16 value) {
      writeContext.getBuffer().writeInt16(value.toBits());
    }

    @Override
    public BFloat16 read(ReadContext readContext) {
      return BFloat16.fromBits(readContext.getBuffer().readInt16());
    }
  }

  public static void registerDefaultSerializers(TypeResolver resolver) {
    // primitive types will be boxed.
    Config config = resolver.getConfig();
    resolver.registerInternalSerializer(
        boolean.class, new BooleanSerializer(config, boolean.class));
    resolver.registerInternalSerializer(byte.class, new ByteSerializer(config, byte.class));
    resolver.registerInternalSerializer(short.class, new ShortSerializer(config, short.class));
    resolver.registerInternalSerializer(char.class, new CharSerializer(config, char.class));
    resolver.registerInternalSerializer(int.class, new IntSerializer(config, int.class));
    resolver.registerInternalSerializer(long.class, new LongSerializer(config, long.class));
    resolver.registerInternalSerializer(float.class, new FloatSerializer(config, float.class));
    resolver.registerInternalSerializer(double.class, new DoubleSerializer(config, double.class));
    resolver.registerInternalSerializer(
        Boolean.class, new BooleanSerializer(config, Boolean.class));
    resolver.registerInternalSerializer(Byte.class, new ByteSerializer(config, Byte.class));
    resolver.registerInternalSerializer(Short.class, new ShortSerializer(config, Short.class));
    resolver.registerInternalSerializer(
        Character.class, new CharSerializer(config, Character.class));
    resolver.registerInternalSerializer(Integer.class, new IntSerializer(config, Integer.class));
    resolver.registerInternalSerializer(Long.class, new LongSerializer(config, Long.class));
    resolver.registerInternalSerializer(Float.class, new FloatSerializer(config, Float.class));
    resolver.registerInternalSerializer(Double.class, new DoubleSerializer(config, Double.class));
    resolver.registerInternalSerializer(
        Float16.class, new Float16Serializer(config, Float16.class));
    resolver.registerInternalSerializer(
        BFloat16.class, new BFloat16Serializer(config, BFloat16.class));
  }
}
